gateway/src/index.js

```diff
@@ -101,47 +104,160 @@ app.all("/:gameId/*", async (request, response) => {
+				break;
 		}
 	}
-	const queryString = Object.keys(query).length
-		? "?" + querystring.stringify(query)
-		: "";
-	try {
-		const { data } = await axios({
-			method: request.method,
-			url: `${gameApi.url}${endpoint}${queryString}`,
-			data: { ...request.body, totp },
-			timeout: 5000,
-		});
-		response.json(data);
-		gameApi.timeoutStartTime = 0;
-	} catch (err) {
-		if (err.code === "ECONNREFUSED") {
-			debugLog("Game API", gameApi.url, err.message);
-			response.status(504).json({ message: "Game API offline" });
-			if (!gameApi.timeoutStartTime) {
-				gameApi.timeoutStartTime = Date.now();
-			} else if (Date.now() - gameApi.timeoutStartTime > 1000 * 60) {
-				delete knownGameApis[gameId];
-			}
-		} else if (err.response && err.response.status) {
+	for (let i = 0; i <= MAX_AUTH_ATTEMPTS; i++)  {
+		if (i === MAX_AUTH_ATTEMPTS) {
 			debugLog(
 				"Game API",
+				gameApi.id,
+				request.method,
 				gameApi.url,
-				err.message,
-				`Status: ${err.response.status} ${err.response.statusText}`,
-				err.response.data
+				"Maximum authentication attempts reached."
 			);
-			response.status(err.response.status).json(err.response.data);
-		} else {
-			console.warn("Game API", err);
-			response.status(500).json({});
+			response.status(502).json({ message: "Failed to authenticate with game API." });
+			break;
+		}
+		try {
+			const queryParamString = Object.keys(queryParams).length
+				? "?" + querystring.stringify(queryParams)
+				: "";
+			const { data } = await axios({
+				method: request.method,
+				url: `${requestURL}${queryParamString}`,
+				data: bodyParams,
+				headers,
+				timeout: 5000,
+			});
+			response.json(data);
+			gameApi.timeoutStartTime = 0;
+			break;
+		} catch (err) {
+			let retryQuery = false;
+			if (err.code === "ECONNREFUSED") {
+				debugLog("Game API", gameApi.url, err.message);
+				response.status(504).json({ message: "Game API offline" });
+				if (!gameApi.timeoutStartTime) {
+					gameApi.timeoutStartTime = Date.now();
+				} else if (Date.now() - gameApi.timeoutStartTime > 1000 * 60) {
+					delete knownGameApis[gameId];
+				}
+			} else if (err.response && err.response.status) {
+				const { status, statusText } = err.response;
+				const responseHeaders = err.response.headers;
+				const responseData = err.response.data;
+				switch(status) {
+					case 401:
+						const authenticate = responseHeaders["WWW-Authenticate"];
+						debugLog(
+							"GameAPI",
+							gameApi.id,
+							request.method,
+							gameApi.url,
+							"Received unauthorized header:",
+							authenticate
+						);
+						const authenticateMap = {}
+						authenticate.split(", ").forEach((kvs) => {
+							let { key, value } = kvs.split("=");
+							if (key.left(7) === "Digest ") {
+								key = key.substring(7);
+							}
+							if (value[0] === '"' && value[value.length - 1] == '"') {
+								value = value.substring(1, value.length - 2);
+							}
+							authenticateMap[key] = value;
+						});
```

Comment:
This (parsing http headers by character) feels like something a library that knows the full spec (and exceptions to the spec) ought to handle, but sure for now if it works. This whole function is now awfully large and should probably be split up so the index.js isn't 60% about auth specifics.

I'll just merge for now though since the whole auth thing is optional anyway.

Reply:
Agreed on both counts. Maybe log an issue for them?